src/persist/src/file.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! In-memory implementations for testing and benchmarking.
+
+use std::collections::{HashMap, HashSet};
+use std::convert::TryInto;
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Seek, SeekFrom, Write as StdWrite};
+use std::sync::{Arc, Mutex};
+
+use crate::error::Error;
+use crate::mem::MemSnapshot;
+use crate::persister::{Meta, Persister, Write};
+use crate::{Id, Token};
+
+/// A naive implementation of [Write] and [Meta] backed by a file.
+#[derive(Clone, Debug)]
+pub struct FileStream {
+    dataz: Arc<Mutex<File>>,
+}
+
+impl FileStream {
+    /// Create a new FileStream
+    pub fn new(name: String) -> Self {
+        // TODO: could not get truncate to work here.
+        let file = OpenOptions::new()
+            .write(true)
+            .append(true)
+            .create(true)
+            .read(true)
+            .open(name)
+            .expect("failed to create file");
+        FileStream {
+            dataz: Arc::new(Mutex::new(file)),
+        }
+    }
+}
+
+impl Write for FileStream {
+    fn write_sync(&mut self, updates: &[((String, String), u64, isize)]) -> Result<(), Error> {
+        let mut buf = Vec::new();
+
+        for ((key, val), ts, diff) in updates.iter() {
+            let key_bytes = key.as_bytes();
+            let key_bytes_len = key_bytes.len() as u64;
+            let val_bytes = val.as_bytes();
+            let val_bytes_len = val_bytes.len() as u64;
+
+            buf.extend(&key_bytes_len.to_le_bytes());
+            buf.extend(key_bytes);
+            buf.extend(&val_bytes_len.to_le_bytes());
+            buf.extend(val_bytes);
+            buf.extend(&ts.to_le_bytes());
+            buf.extend(&(*diff as i64).to_le_bytes());
+        }
+        let mut guard = self.dataz.lock()?;
+        guard.write_all(&buf).expect("failed to write file");
```

Comment:
This should be returning the error instead of expect, so we can (once we get to this point) return this error message to the user who did the INSERT (or put it in the error stream for a source we're persisting).